### PR TITLE
enhance js unit test coverage

### PR DESCRIPTION
**Pull request by alexbowen**
paying particular atteniton to

- interaction with instant search client
- location search box
- general improvements

the intention of this PR is purely unit tests, DOM interaction tests could be another task as this is sizeable already (although low impact as this functionality is disabled)

## Jira ticket URL

https://dfedigital.atlassian.net/browse/TEVA-897

### CODE REVIEW

File: app/frontend/src/search/ui/input/keyword.test.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    const client = {};
+    let performSearch = null;
+
+    beforeEach(() => {
+        client.refresh = jest.fn();
+        performSearch = jest.spyOn(client, 'refresh');
+    });
+
+    describe('onSubmit', () => {
+        test('does not performs search if unsuccesful', () => {
```
Comment (david-mears-dfe) [inline app/frontend/src/search/ui/input/keyword.test.js:14]: performs -> perform
unsuccesful -> unsuccessful

File: app/frontend/src/search/ui/input/keyword.test.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+            onSubmit(client);
+            expect(performSearch).toHaveBeenCalledTimes(1);
```
Comment (david-mears-dfe) [inline app/frontend/src/search/ui/input/keyword.test.js:16]: Confused that the test description says 'does not perform search' but we expect the search to have been performed once.

File: app/frontend/src/search/ui/input/location.test.js [hunk 3]
```diff
@@ -44,9 +46,10 @@ describe('location', () => {
         });
     });
 
-    describe('getGeolocatedCoordinates', () => {
-        test('does not get called if location supplied that exists in predefined list', () => {
+    describe('onSubmit getGeolocatedCoordinates', () => {
+        test('does not get called if location supplied that exists in predefined list and radius diabled', () => {
```
Comment (david-mears-dfe) [inline app/frontend/src/search/ui/input/location.test.js:50]: diabled -> disabled

File: app/frontend/src/search/ui/input/location.test.js [hunk 4]
```diff
@@ -58,9 +61,10 @@ describe('location', () => {
             expect(performSearch).not.toHaveBeenCalled();
         });
 
-        test('performs a search if succesful', () => {
+        test('performs a search if succesful and radius is enabled', () => {
```
Comment (david-mears-dfe) [inline app/frontend/src/search/ui/input/location.test.js:64]: succesful -> successful

**Comment by david-mears-dfe** [comment]
I've only looked at the second commit of these, so far.

File: app/frontend/src/search/index.js [hunk 5]
```diff
@@ -11,139 +9,118 @@ import { templates, renderContent } from './hits';
...
+        }
+    }),
+    locationSearchBox,
+    keywordSearchBox,
+    locationRadius({
+        container: document.querySelector('#location-radius-select'),
+        attribute: '_geoloc',
+        inputElement: document.getElementById('radius'),
+        onSelection: location => {
+            updateUrlQueryParams('radius', location, window.location.href);
```
Comment (cpjmcquillan) [inline app/frontend/src/search/index.js:79]: The location variable here seems misleading/confusing, could you use `radiusInMiles` instead?

File: app/frontend/src/search/ui/input/location.js [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+            geocodeSuccess(coords, client);
+        });
+    } else {
+        disableRadiusSelect();
+        removeGeocodeAttributes();
+        client.refresh();
+    }
+};
+
+export const shouldGeocode = (query, locations) => stringMatchesPostcode(query) || (query.length && locations.indexOf(query.toLowerCase()) === -1);
```
Comment (cpjmcquillan) [inline app/frontend/src/search/ui/input/location.js:18]: To me this logic would be simpler if the function was actually `shouldNotGeocode`. Then we only need to check the presence of the location in the list.

File: app/frontend/src/search/ui/alert.js [hunk 7]
```diff
@@ -0,0 +1,43 @@
+import { extractQueryParams } from '../../lib/utils';
+import '../../polyfill/remove.polyfill';
+
+const JOB_ALERT_URL = '/subscriptions/new';
+
+export const getJobAlertLink = url => {
```
Comment (cpjmcquillan) [inline app/frontend/src/search/ui/alert.js:6]: Just a note that some aspect of this isn't always working as intended https://dfedigital.atlassian.net/browse/TEVA-882

File: app/frontend/src/search/ui/input/location.test.js [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    beforeEach(() => {
+        jest.resetAllMocks();
+
+        client.refresh = jest.fn();
+        performSearch = jest.spyOn(client, 'refresh');
+
+        getGeolocatedCoordinates.mockReturnValue(Promise.resolve({ success: true }));
+    });
+
+    describe('shouldGeocode', () => {
```
Comment (cpjmcquillan) [inline app/frontend/src/search/ui/input/location.test.js:24]: `shouldNotGeocode`

File: app/frontend/src/search/ui/input/location.test.js [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+        test('returns true if location is not in predefined list or matches a postcode pattern', () => {
```
Comment (cpjmcquillan) [inline app/frontend/src/search/ui/input/location.test.js:25]: `returns false`

File: app/frontend/src/search/ui/input/location.test.js [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+            expect(shouldNotGeocode('harlow', locations)).toBe(false);
+            expect(shouldNotGeocode('SE17 4BT', locations)).toBe(false);
+        });
+
+        test('returns false if location is in predefined list', () => {
```
Comment (cpjmcquillan) [inline app/frontend/src/search/ui/input/location.test.js:30]: `returns true`

File: app/frontend/src/search/ui/input/location.test.js [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+    describe('onSubmit getGeolocatedCoordinates', () => {
+        test('does not get called if location supplied that exists in predefined list and radius disabled', () => {
+            onSubmit('london', locations, client);
+            expect(disableRadiusSelect).toHaveBeenCalledTimes(1);
+            expect(getGeolocatedCoordinates).not.toHaveBeenCalled();
+        });
+    });
+
+    describe('geocodeSuccess', () => {
+        test('does not performs search if unsuccesful', () => {
```
Comment (cpjmcquillan) [inline app/frontend/src/search/ui/input/location.test.js:58]: `perform`

File: app/frontend/src/search/ui/input/radius.js [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+            widgetParams.onSelection(event.target.value);
+        });
+        disableRadiusSelect();
+    }
+
+    query ? updateUrlQueryParams(widgetParams.key, query) : false;
+};
+
+export const enableRadiusSelect = () => {
+
```
Comment (cpjmcquillan) [inline app/frontend/src/search/ui/input/radius.js:17]: empty line

File: app/frontend/src/search/ui/alert.js (hunk 7 above)
Comment (alexbowen) [reply app/frontend/src/search/ui/alert.js:6]: sure i am aware of this ticket, i dont want to neccessarily try fixing anything already known like this as part of this PR

**Pull request merged**